src/swarm-tui/Swarm/TUI/Controller/EventHandlers.hs

```diff
@@ -0,0 +1,131 @@
+-- 5. Finally in 'Swarm.TUI.Controller' the Brick event handler calls the stored dispatchers.
+
+-- | Create key dispatchers that call (abstract) event handlers based on given key config.
+--
+-- Fails if any key events have conflict within one dispatcher or when a main dispatcher
+-- has conflict with one of the subdispatchers.
+createKeyDispatchers ::
+  (Has (Throw SystemFailure) sig m) =>
+  KeyConfig SwarmEvent ->
+  m SwarmKeyDispatchers
+createKeyDispatchers config = do
+  mainGameDispatcher <- buildDispatcher mainEventHandlers
+  let buildSubMainDispatcher = buildSubDispatcher "Main game events" mainGameDispatcher
+  replDispatcher <- buildSubMainDispatcher "REPL panel events" replEventHandlers
+  worldDispatcher <- buildSubMainDispatcher "World view panel events" worldEventHandlers
+  robotDispatcher <- buildSubMainDispatcher "Robot inventory panel events" robotEventHandlers
+  return SwarmKeyDispatchers {..}
+ where
+  -- this error handling code is modified version of the brick demo app:
+  -- https://github.com/jtdaugherty/brick/blob/764e66897/programs/CustomKeybindingDemo.hs#L216
+  buildDispatcher handlers = case keyDispatcher config handlers of
+    Left collisions ->
+      throwLoadingFailure $
+        "Error: some key events have the same keys bound to them.\n"
+          : handlerErrors collisions
+    Right d -> return d
+  buildSubDispatcher parentName parentDispatcher name handlers = do
+    d <- buildDispatcher handlers
+    let collisions = conflicts parentDispatcher d
+    if null collisions
+      then return d
+      else
+        throwLoadingFailure $
+          ("Error: some key events have keys bound to them in '" <> parentName <> "' and in '" <> name <> "'")
+            : handlerErrors collisions
+
+  throwLoadingFailure = throwError . CustomFailure . T.intercalate "\n"
+  handlerErrors collisions = flip map collisions $ \(b, hs) ->
+    let hsm = "Handlers with the '" <> BK.ppBinding b <> "' binding:"
+        hss = flip map hs $ \h ->
+          let trigger = case BK.kehEventTrigger $ BK.khHandler h of
+                ByKey k -> "triggered by the key '" <> BK.ppBinding k <> "'"
+                ByEvent e -> "triggered by the event '" <> fromMaybe "<unknown>" (BK.keyEventName swarmEvents e) <> "'"
+              desc = BK.handlerDescription $ BK.kehHandler $ BK.khHandler h
+           in "  " <> desc <> " (" <> trigger <> ")"
+     in T.intercalate "\n" (hsm : hss)
+
+-- | Take two dispatchers (that do not have conflict themselves) and find conflicting keys between them.
+conflicts :: SwarmKeyDispatcher -> SwarmKeyDispatcher -> [(Binding, [KeyHandler SwarmEvent (EventM Name AppState)])]
+conflicts d1 d2 = combine <$> badGroups
+ where
+  l1 = keyDispatcherToList d1
+  l2 = keyDispatcherToList d2
+  gs = groupBy ((==) `on` fst) $ sortBy (compare `on` fst) (l1 <> l2)
+  badGroups = filter ((1 <) . length) $ mapMaybe NE.nonEmpty gs
```

Comment:
Use the `NonEmpty` version of [`groupWith`](https://hackage.haskell.org/package/base-4.20.0.1/docs/Data-List-NonEmpty.html#v:groupWith)?
```suggestion
  gs = NE.groupWith fst $ sortBy (compare `on` fst) (l1 <> l2)
  badGroups = filter ((1 <) . length) gs
```

Reply:
Cool, this is even nicer than `groupOn` from `Extra`.